README.md
# JLed - Advanced LED Library

![run tests](https://github.com/jandelgado/jled/workflows/run%20tests/badge.svg)
[![Coverage Status](https://coveralls.io/repos/github/jandelgado/jled/badge.svg?branch=master&dummy=1)](https://coveralls.io/github/jandelgado/jled?branch=master)

An embedded C++ library to control LEDs. It uses a **non-blocking** approach and can
control LEDs in simple (**on**/**off**) and complex (**blinking**,
**breathing** and more) ways in a **time-driven** manner.

JLed got some [coverage on Hackaday](https://hackaday.com/2018/06/13/simplifying-basic-led-effects/)
and someone did a [video tutorial for JLed](https://youtu.be/x5V2vdpZq1w)  - Thanks!

JLed in action | Interactive JLed playground
:-------------:|:--------------------------------------------------:
<a href="examples/multiled"><img alt="breathing, blinking, fadeon and -off at the same time" height=200 src="doc/jled.gif"></a>|<a href="https://jandelgado.github.io/jled-wasm"><img alt="jled running in the browser" height=200  src="doc/jled-wasm.png"></a>

## Example

```c++
// breathe LED (on gpio 9) 6 times for 1500ms, waiting for 500ms after each run
#include <jled.h>

auto led_breathe = JLed(9).Breathe(1500).Repeat(6).DelayAfter(500);

void setup() { }

void loop() {
  led_breathe.Update();
}
```

## Contents

<!-- vim-markdown-toc GFM -->

* [Features](#features)
* [Cheat Sheet](#cheat-sheet)
* [Installation](#installation)
    * [Arduino IDE](#arduino-ide)
    * [PlatformIO](#platformio)
* [Usage](#usage)
    * [Effects](#effects)
        * [Static on and off](#static-on-and-off)
            * [Static on example](#static-on-example)
        * [Blinking](#blinking)
            * [Blinking example](#blinking-example)
        * [Breathing](#breathing)
            * [Breathing example](#breathing-example)
        * [Candle](#candle)
            * [Candle example](#candle-example)
        * [FadeOn](#fadeon)
            * [FadeOn example](#fadeon-example)
        * [FadeOff](#fadeoff)
        * [User provided brightness function](#user-provided-brightness-function)
            * [User provided brightness function example](#user-provided-brightness-function-example)
        * [Delays and repetitions](#delays-and-repetitions)
            * [Initial delay before effect starts](#initial-delay-before-effect-starts)
            * [Delay after effect finished](#delay-after-effect-finished)
            * [Repetitions](#repetitions)
        * [State functions](#state-functions)
            * [Update](#update)
            * [UpdateAndFinally](#updateandfinally)
            * [IsRunning](#isrunning)
            * [Reset](#reset)
            * [Immediate Stop](#immediate-stop)
        * [Misc functions](#misc-functions)
            * [Low active for inverted output](#low-active-for-inverted-output)
            * [Maximum brightness level](#maximum-brightness-level)
    * [Controlling a group of LEDs](#controlling-a-group-of-leds)
* [Framework notes](#framework-notes)
* [Platform notes](#platform-notes)
    * [ESP8266](#esp8266)
    * [ESP32](#esp32)
    * [STM32](#stm32)
        * [Arduino framework](#arduino-framework)
    * [Raspberry Pi Pico](#raspberry-pi-pico)
* [Example sketches](#example-sketches)
    * [PlatformIO](#platformio-1)
    * [Arduino IDE](#arduino-ide-1)
* [Extending](#extending)
    * [Support new hardware](#support-new-hardware)
* [Unit tests](#unit-tests)
* [Contributing](#contributing)
* [FAQ](#faq)
    * [How do I check if a JLed object is still being updated?](#how-do-i-check-if-a-jled-object-is-still-being-updated)
    * [How do I restart an effect?](#how-do-i-restart-an-effect)
    * [How do I change a running effect?](#how-do-i-change-a-running-effect)
* [Author and Copyright](#author-and-copyright)
* [License](#license)

<!-- vim-markdown-toc -->

## Features

* non-blocking
* effects: simple on/off, breathe, blink, candle, fade-on, fade-off, [user-defined](examples/morse) (e.g. morse)
* supports inverted  polarity of LED
* easy configuration using fluent interface
* can control groups of LEDs sequentially or in parallel
* Portable: Arduino, ESP8266, ESP32, Mbed, Raspberry Pi Pico and more platforms
  compatible, runs even in the [browser](https://jandelgado.github.io/jled-wasm)
* supports Arduino, [mbed](https://www.mbed.com) and Raspberry Pi Pico SDKs
* well [tested](https://coveralls.io/github/jandelgado/jled)

## Cheat Sheet

![JLed Cheat Sheet](doc/cheat_sheet.jpg)

## Installation

### Arduino IDE

In the main menu of the Arduino IDE, select `Sketch` > `Include Library` >
`Manage Libraries...` and search for `jled`, then press `install`.

### PlatformIO

Add `jled` to your library dependencies in your `platformio.ini` project file,
e.g.

```ini
...
[env:nanoatmega328]
platform = atmelavr
board = nanoatmega328
framework = arduino
lib_deps=jled
...
```

## Usage

First, the LED object is constructed and configured, then the state is updated
with subsequent calls to the `Update()` method, typically from the `loop()`
function. While the effect is active, `Update` returns `true`, otherwise
`false`.

The constructor takes the pin, to which the LED is connected to as
the only argument. Further configuration of the LED object is done using a fluent
interface, e.g. `JLed led = JLed(13).Breathe(2000).DelayAfter(1000).Repeat(5)`.
See the examples section below for further details.

### Effects

#### Static on and off

Calling `On()` turns the LED on.  To immediately turn a LED on, make a call
like `JLed(LED_BUILTIN).On().Update()`.

`Off()` works like `On()`, except that it turns the LED off, i.e. it sets the
brightness to 0.

Use the `Set(uint8_t brightness)` method to set the brightness to the given
value, i.e. `Set(255)` is equivalent to calling `On()` and `Set(0)` is
equivalent to calling `Off()`.

Technically `Set`,  `On` and `Off` are effects with a period of 1ms that 
set the brightness to a constant value.

##### Static on example

```c++
#include <jled.h>

// turn builtin LED on after 1 second.
auto led = JLed(LED_BUILTIN).On().DelayBefore(1000);

void setup() { }

void loop() {
  led.Update();
}
```

#### Blinking

In blinking mode, the LED cycles through a given number of on-off cycles, on-
and off-cycle durations are specified independently. The `Blink()` method takes
the duration for the on- and off cycle as arguments.

##### Blinking example

```c++
#include <jled.h>

// blink internal LED every second; 1 second on, 0.5 second off.
auto led = JLed(LED_BUILTIN).Blink(1000, 500).Forever();

void setup() { }

void loop() {
  led.Update();
}
```

#### Breathing

In breathing mode, the LED smoothly changes the brightness using PWM. The
`Breathe()` method takes the period of the effect as an argument.

##### Breathing example

```c++
#include <jled.h>

// connect LED to pin 13 (PWM capable). LED will breathe with period of
// 2000ms and a delay of 1000ms after each period.
auto led = JLed(13).Breathe(2000).DelayAfter(1000).Forever();

void setup() { }

void loop() {
  led.Update();
}
```

It is also possible to specify fade-on, on and fade-off durations for the breathing mode to customize the effect.

```c++
// LED will fade-on in 500ms, stay on for 1000ms, and fade-off in 500ms.
// It will delay for 1000ms afterwards and continue the pattern.
auto led = JLed(13).Breathe(500, 1000, 500).DelayAfter(1000).Forever();
```

#### Candle

In candle mode, the random flickering of a candle or fire is simulated. 
The builder method has the following signature:
  `Candle(uint8_t speed, uint8_t jitter, uin16_t period)`

* `speed` - controls the speed of the effect. 0 for fastest, increasing speed 
  divides into halve per increment. The default value is 7.
* `jitter` - the amount of jittering. 0 none (constant on), 255 maximum. Default
  value is 15.
* `period` - Period of effect in ms.  The default value is 65535 ms.

The default settings simulate a candle. For a fire effect for example use
call the method with `Candle(5 /*speed*/, 100 /* jitter*/)`. 

##### Candle example

```c++
#include <jled.h>

// Candle on LED pin 13 (PWM capable). 
auto led = JLed(13).Candle();

void setup() { }

void loop() {
  led.Update();
}
```

#### FadeOn

In FadeOn mode, the LED is smoothly faded on to 100% brightness using PWM. The
`FadeOn()` method takes the period of the effect as an argument.

The brightness function uses an approximation of this function (example with
period 1000):

[![fadeon function](doc/fadeon_plot.png)](https://www.wolframalpha.com/input/?i=plot+(exp(sin((t-1000%2F2.)*PI%2F1000))-0.36787944)*108.0++t%3D0+to+1000)


##### FadeOn example

```c++
#include <jled.h>

// LED is connected to pin 9 (PWM capable) gpio
auto led = JLed(9).FadeOn(1000).DelayBefore(2000);

void setup() { }

void loop() {
  led.Update();
}
```

#### FadeOff

In FadeOff mode, the LED is smoothly faded off using PWM. The fade starts at
100% brightness. Internally it is implemented as a mirrored version of the
FadeOn function, i.e. FadeOn(t) = FadeOff(period-t).  The `FadeOff()` method
takes the period of the effect as argument.

#### User provided brightness function

It is also possible to provide a user defined brightness evaluator. The class
must be derived from the `jled::BrightnessEvaluator` class and implement
two methods:

* `uint8_t Eval(uint32_t t) const` - the brightness evaluation function that
  calculates a brightness for the given time `t`. The brightness must be returned
  as an unsigned byte, where 0 means LED off and 255 means full brightness.
* `uint16_t Period() const` - period of the effect.

All time values are specified in milliseconds. 

The [user_func](examples/user_func) example demonstrates a simple user provided
brightness function, while the [morse](examples/morse) example shows how a more
complex application, allowing you to send morse codes (not necessarily with an
LED), can be realized.

##### User provided brightness function example

The example uses a user provided function to calculate the brightness.

```c++
class UserEffect : public jled::BrightnessEvaluator {
  public:
    uint8_t Eval(uint32_t t) const override {
        // this function changes between 0 and 255 and
        // vice versa every 250 ms.
        return 255*((t/250)%2);
    }
    // duration of effect: 5 seconds.

CHANGELOG.md
# JLed changelog (github.com/jandelgado/jled)

## [2021-12-12] 4.9.0

* new: `JLed::UpdateAndFinally` method allowing to trigger custom code after
       an effect is finished. See [README.md#updateandfinally] for details.

## [2021-10-18] 4.8.0

* new: make `Breathe` method more flexible (#78, thanks to @boraozgen)

## [2021-10-13] 4.7.1

* fix: correct handling of time rollover (#80, thanks to @boraozgen)

## [2021-02-02] 4.7.0

* new: support for Raspberry Pi Pico added

## [2021-02-02] 4.6.1

* fix: `Forever()` on sequence had no effect (#68)

## [2021-01-24] 4.6.0

* new: JLedSequence can be configured to play the sequence multiple times
       using the `Repeat()` and `Forever()` methods
* drop travis-ci, use github actions

## [2020-10-24] 4.5.2

* fix: ESP32 led glimming when using low-active connection (#60)

## [2020-07-01] 4.5.1

* fix: support for Nano 33 BLE (#53)

## [2020-02-23] 4.5.0

* new: `JLed::MaxBrightness(uint8_t level)` method to limit output of effects
  (implements #43).

## [2020-02-21] 4.4.0

* JLed now supports the mbed framework. See README.md and `examples/multiled_mbed`
  for examples.

## [2019-09-21] 4.3.0

* new example: [custom HAL](examples/custom_hal/custom_hal.ino) showing
  how to implment a custom HAL.

## [2019-08-30] 4.2.1

* fix: make sure memory alignment is correct (caused hard fault on 
  SAMD21). Fixes #27.

## [2019-06-20] 4.2.0

* changing an effect resets the Jled object so it starts over with the 
  new effect (see #25). Prior to this change, calling `Reset()` was
  necessary.

## [2019-05-11] 4.1.2

* fix: ESP32 dynamic channel assignment fixed. Sequence demo now working
       as expected (see #22).

## [2019-05-07] 4.1.1

* fix: version format in library.properties (removed leading `v`; see #21)

## [2019-03-10] v4.1.0

* change: clean up interface and simplify code: `On()` no longer takes an
  optional brightness argument. Call `Set(uint8_t brightness)` instead.
* documentation update

## [2019-03-10] v4.0.0

In addition to the changes introduced with `v4.0.0-rc0` and `v4.0.0-rc1`, the
`v4.0.0` relases adds/changes the following:

### Added

* new `Candle()` effect added for candles and fire like effects

### Changed

* The user provided brightness class no longer needs a `Clone()` method. See
  [example](examples/user_func/user_func.ino) for an example

### Fixed

* Makefile (unit testing) dependency checking fixed

## [2019-02-17] v4.0.0-rc1

### Changed

* fix: byte buffer alignment for ESP8266 set to DWORD boundary making ESP8266
  run again with JLed 4.x
* arduino HAL now does lazy call to pinMode() to prevent STM32 problems
* simplified morse example code

## [2019-01-23] v4.0.0-rc0

### Added

* `JLed::Reset()` - resets the led to it's initial state allowing to 
  to start over
* `JLed::IsRunning()` - return true if effect is active, else false
* new class `JLedSequence` to update JLed objects simultanously or 
  sequentially. See [README](README.md#controlling-a-group-of-leds) for details.
* added new [morse example](examples/morse)
* clean separation between hardware specific and common code, making
  extendability easy
* added STM32 example to [platformio.ini](platformio.ini)

### Changed

* the brightness user function pointer was replaced by an object of type
  BrightnessEvaluator. Migration of code should be straight forward, see
  below

#### old brightness function 

In JLed version prio to version 4.0.0, a function pointer was used to specify
a user provided brightness function.

```c++
// this function returns changes between 0 and 255 and vice versa every 250 ms.
uint8_t blinkFunc(uint32_t t, uint16_t, uintptr_t) {
  return 255*((t/250)%2);
}

// Run blinkUserFunc for 5000ms
JLed led = JLed(LED_BUILTIN).UserFunc(blinkFunc, 5000);
```

#### new BrightnessEvaluator class

The user function is replaced by a class, which provides more flexibility:

```c++
class UserEffect : public jled::BrightnessEvaluator {
    uint8_t Eval(uint32_t t) const {
        // this function returns changes between 0 and 255 and
        // vice versa every 250 ms.
        return 255*((t/250)%2);
    }
    uint16_t Period() const { return 5000; }
};

UserEffect userEffect;
JLed led = JLed(LED_BUILTIN).UserFunc(&userEffect);
```

### Removed

* `JLed::Invert()` method was removed since became redundant with LowActive()


## [2018-10-03] v3.0.0

* Major refactoring making support of different platforms easier
* ESP32 support added
* Unit tests refactored

## [2018-09-22] v2.4.0

* `JLed::Update()` now returns a `bool` indicating if the effect is still
  active (true), or finished (false).

## [2018-09-22] v2.3.0

* ESP8266 platform: scaling from 8 to 10 bit improved. The scaling makes sure
  that 0 is mapped to 0 and 255 is mapped to 1023, preserving min/max 
  relationships in both ranges.

## [2018-06-09] v2.2.3

### Fixes

* ESP8266 platform: analogWrite() resoultion of 10 bit is now honoured.
  Previously only a range of 0..255 was used, which resulted in output being
  dimmed.

### Added

* It's never to late for a changelog ;)
* ESP8266 environment added to [platform.ini](platform.ini)


src/jled_base.h
// Copyright (c) 2017-2021 Jan Delgado <jdelgado[at]gmx.net>
// https://github.com/jandelgado/jled
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to
// deal in the Software without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
// sell copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
// IN THE SOFTWARE.
//
#ifndef SRC_JLED_BASE_H_
#define SRC_JLED_BASE_H_

#include <inttypes.h>  // types, e.g. uint8_t
#include <stddef.h>    // size_t

// JLed - non-blocking LED abstraction library.
//
// Example Arduino sketch:
//   auto led = JLed(LED_BUILTIN).Blink(500, 500).Repeat(10).DelayBefore(1000);
//
//   void setup() {}
//
//   void loop() {
//     led.Update();
//   }

namespace jled {

static constexpr uint8_t kFullBrightness = 255;
static constexpr uint8_t kZeroBrightness = 0;

uint8_t fadeon_func(uint32_t t, uint16_t period);
uint8_t rand8();
void rand_seed(uint32_t s);
uint8_t scale5(uint8_t val, uint8_t factor);

// a function f(t,period,param) that calculates the LEDs brightness for a given
// point in time and the given period. param is an optionally user provided
// parameter. t will always be in range [0..period-1].
// f(period-1,period,param) will be called last to calculate the final state of
// the LED.
class BrightnessEvaluator {
 public:
    virtual uint16_t Period() const = 0;
    virtual uint8_t Eval(uint32_t t) const = 0;
};

class CloneableBrightnessEvaluator : public BrightnessEvaluator {
 public:
    virtual BrightnessEvaluator* clone(void* ptr) const = 0;
    static void* operator new(size_t, void* ptr) { return ptr; }
    static void operator delete(void*) {}
};

class ConstantBrightnessEvaluator : public CloneableBrightnessEvaluator {
    uint8_t val_;

 public:
    ConstantBrightnessEvaluator() = delete;
    explicit ConstantBrightnessEvaluator(uint8_t val) : val_(val) {}
    BrightnessEvaluator* clone(void* ptr) const override {
        return new (ptr) ConstantBrightnessEvaluator(*this);
    }
    uint16_t Period() const override { return 1; }
    uint8_t Eval(uint32_t) const override { return val_; }
};

// BlinkBrightnessEvaluator does one on-off cycle in the specified period
class BlinkBrightnessEvaluator : public CloneableBrightnessEvaluator {
    uint16_t duration_on_, duration_off_;

 public:
    BlinkBrightnessEvaluator() = delete;
    BlinkBrightnessEvaluator(uint16_t duration_on, uint16_t duration_off)
        : duration_on_(duration_on), duration_off_(duration_off) {}
    BrightnessEvaluator* clone(void* ptr) const override {
        return new (ptr) BlinkBrightnessEvaluator(*this);
    }
    uint16_t Period() const override { return duration_on_ + duration_off_; }
    uint8_t Eval(uint32_t t) const override {
        return (t < duration_on_) ? kFullBrightness : kZeroBrightness;
    }
};

// fade LED on
class FadeOnBrightnessEvaluator : public CloneableBrightnessEvaluator {
    uint16_t period_;

 public:
    FadeOnBrightnessEvaluator() = delete;
    explicit FadeOnBrightnessEvaluator(uint16_t period) : period_(period) {}
    BrightnessEvaluator* clone(void* ptr) const override {
        return new (ptr) FadeOnBrightnessEvaluator(*this);
    }
    uint16_t Period() const override { return period_; }
    uint8_t Eval(uint32_t t) const override { return fadeon_func(t, period_); }
};

// fade LED off
class FadeOffBrightnessEvaluator : public CloneableBrightnessEvaluator {
    uint16_t period_;

 public:
    FadeOffBrightnessEvaluator() = delete;
    explicit FadeOffBrightnessEvaluator(uint16_t period) : period_(period) {}
    BrightnessEvaluator* clone(void* ptr) const override {
        return new (ptr) FadeOffBrightnessEvaluator(*this);
    }
    uint16_t Period() const override { return period_; }
    uint8_t Eval(uint32_t t) const override {
        return fadeon_func(period_ - t, period_);
    }
};

// The breathe func is composed by fade-on, on and fade-off phases. For fading
// we approximate the following function:
//   y(x) = exp(sin((t-period/4.) * 2. * PI / period)) - 0.36787944) *  108.)
// idea see:
//   http://sean.voisen.org/blog/2011/10/breathing-led-with-arduino/
// But we do it with integers only.
class BreatheBrightnessEvaluator : public CloneableBrightnessEvaluator {
    uint16_t duration_fade_on_;
    uint16_t duration_on_;
    uint16_t duration_fade_off_;

 public:
    BreatheBrightnessEvaluator() = delete;
    explicit BreatheBrightnessEvaluator(uint16_t duration_fade_on,
                                        uint16_t duration_on,
                                        uint16_t duration_fade_off)
        : duration_fade_on_(duration_fade_on),
          duration_on_(duration_on),
          duration_fade_off_(duration_fade_off) {}
    BrightnessEvaluator* clone(void* ptr) const override {
        return new (ptr) BreatheBrightnessEvaluator(*this);
    }
    uint16_t Period() const override {
        return duration_fade_on_ + duration_on_ + duration_fade_off_;
    }
    uint8_t Eval(uint32_t t) const override {
        if (t < duration_fade_on_)
            return fadeon_func(t, duration_fade_on_);
        else if (t < duration_fade_on_ + duration_on_)
            return kFullBrightness;
        else if (t < Period())
            return fadeon_func(Period() - t, duration_fade_off_);
        else
            return kZeroBrightness;
    }
};

class CandleBrightnessEvaluator : public CloneableBrightnessEvaluator {
    uint8_t speed_;
    uint8_t jitter_;
    uint16_t period_;
    mutable uint8_t last_ = 5;
    mutable uint32_t last_t_ = 0;

 public:
    CandleBrightnessEvaluator() = delete;

    // speed - speed of effect (0..15). 0 fastest. Each increment by 1
    //         halfes the speed.
    // jitter - amount of jittering to apply. 0 - no jitter, 15 - candle,
    //                                        64 - fire, 255 - storm
    CandleBrightnessEvaluator(uint8_t speed, uint8_t jitter, uint16_t period)
        : speed_(speed), jitter_(jitter), period_(period) {}

    BrightnessEvaluator* clone(void* ptr) const override {
        return new (ptr) CandleBrightnessEvaluator(*this);
    }

    uint16_t Period() const override { return period_; }
    uint8_t Eval(uint32_t t) const override {
        // idea from
        // https://cpldcpu.wordpress.com/2013/12/08/hacking-a-candleflicker-led/
        // TODO(jd) finetune values
        static constexpr uint8_t kCandleTable[] = {
            5, 10, 20, 30, 50, 20, 30, 40, 50, 60, 70, 80, 90, 100, 110, 255};
        if ((t >> speed_) == last_t_) return last_;
        last_t_ = (t >> speed_);
        const auto rnd = rand8() & 255;
        last_ = (rnd >= jitter_) ? 255 : (50 + kCandleTable[rnd & 0xf]);
        return last_;
    }
};

template <typename HalType, typename B>
class TJLed {
 protected:
    // pointer to a (user defined) brightness evaluator.
    BrightnessEvaluator* brightness_eval_ = nullptr;
    // Hardware abstraction giving access to the MCU
    HalType hal_;

    void Write(uint8_t val) {
        val = scale5(val, maxBrightness_);
        hal_.analogWrite(IsLowActive() ? kFullBrightness - val : val);
    }

 public:
    TJLed() = delete;
    explicit TJLed(const HalType& hal)
        : hal_{hal},
          state_{ST_RUNNING},
          bLowActive_{false},
          maxBrightness_{(1 << kBitsBrightness) - 1} {}

    explicit TJLed(typename HalType::PinType pin) : TJLed{HalType{pin}} {}

    TJLed(const TJLed& rLed) : hal_{rLed.hal_} { *this = rLed; }

    B& operator=(const TJLed<HalType, B>& rLed) {
        state_ = rLed.state_;
        bLowActive_ = rLed.bLowActive_;
        maxBrightness_ = rLed.maxBrightness_;
        num_repetitions_ = rLed.num_repetitions_;
        last_update_time_ = rLed.last_update_time_;
        delay_before_ = rLed.delay_before_;
        delay_after_ = rLed.delay_after_;
        time_start_ = rLed.time_start_;
        hal_ = rLed.hal_;

        if (rLed.brightness_eval_ !=
            reinterpret_cast<const BrightnessEvaluator*>(
                rLed.brightness_eval_buf_)) {
            // nullptr or points to (external) user provided evaluator
            brightness_eval_ = rLed.brightness_eval_;
        } else {
            brightness_eval_ =
                (reinterpret_cast<const CloneableBrightnessEvaluator*>(
                     rLed.brightness_eval_))
                    ->clone(brightness_eval_buf_);
        }
        return static_cast<B&>(*this);
    }

    HalType& Hal() { return hal_; }

    // The Callback declartion for function to be called when an effect is
    // finished.  We can not use std::function<void(B& led)> f) here, since the
    // functional header is not available for AVR. That means we can only use
    // lambdas without captures. As a workaround, we allow an additional void*
    // pointer to be passed, which can point to arbitrary data.
    typedef void (*FinallyCallback)(B& led, void* data);    // NOLINT

    // Call Update and when the effect is finished, call the provided function
    // f, which gets a reference to this JLed object and an optional data
    // pointer.  Use e.g. to reconfigure & restart after an effect was played.
    bool UpdateAndFinally(FinallyCallback f, void* data = nullptr) {
        if (!IsRunning()) {
            return false;
        }
        if (!Update()) {
            f(static_cast<B&>(*this), data);
            return false;
        }
        return true;
    }

    // Update the effect, return true if the effect is running, false if the
    // effect finished.
    bool Update() { return Update(hal_.millis()); }

    // Set physical LED polarity to be low active. This inverts every
    // signal physically output to a pin.
    B& LowActive() {
        bLowActive_ = true;
        return static_cast<B&>(*this);
    }

    bool IsLowActive() const { return bLowActive_; }

    // turn LED on
    B& On() { return Set(kFullBrightness); }

    // turn LED off
    B& Off() { return Set(kZeroBrightness); }

    // Sets LED to given brightness
    B& Set(uint8_t brightness) {
        // note: we use placement new and therefore not need to keep track of
        // mem allocated
        return SetBrightnessEval(new (brightness_eval_buf_)
                                     ConstantBrightnessEvaluator(brightness));
    }

    // Fade LED on
    B& FadeOn(uint16_t duration) {
        return SetBrightnessEval(new (brightness_eval_buf_)
                                     FadeOnBrightnessEvaluator(duration));
    }

    // Fade LED off - acutally is just inverted version of FadeOn()
    B& FadeOff(uint16_t duration) {
        return SetBrightnessEval(new (brightness_eval_buf_)
                                     FadeOffBrightnessEvaluator(duration));
    }

    // Set effect to Breathe, with the given period time in ms.
    B& Breathe(uint16_t period) { return Breathe(period / 2, 0, period / 2); }

    // Set effect to Breathe, with the given fade on-, on- and fade off-
    // duration values.
    B& Breathe(uint16_t duration_fade_on, uint16_t duration_on,
               uint16_t duration_fade_off) {
        return SetBrightnessEval(
            new (brightness_eval_buf_) BreatheBrightnessEvaluator(
                duration_fade_on, duration_on, duration_fade_off));
    }

    // Set effect to Blink, with the given on- and off- duration values.
    B& Blink(uint16_t duration_on, uint16_t duration_off) {
        return SetBrightnessEval(
            new (brightness_eval_buf_)
                BlinkBrightnessEvaluator(duration_on, duration_off));
    }

    // Set effect to Candle light simulation
    B& Candle(uint8_t speed = 6, uint8_t jitter = 15,
              uint16_t period = 0xffff) {
        return SetBrightnessEval(
            new (brightness_eval_buf_)
                CandleBrightnessEvaluator(speed, jitter, period));
    }

    // Use a user provided brightness evaluator.
    B& UserFunc(BrightnessEvaluator* user_eval) {
        return SetBrightnessEval(user_eval);
    }

    // set number of repetitions for effect.
    B& Repeat(uint16_t num_repetitions) {
        num_repetitions_ = num_repetitions;
        return static_cast<B&>(*this);
    }

    // repeat Forever
    B& Forever() { return Repeat(kRepeatForever); }
    bool IsForever() const { return num_repetitions_ == kRepeatForever; }

    // Set amount of time to initially wait before effect starts. Time is
    // relative to first call of Update() method and specified in ms.
    B& DelayBefore(uint16_t delay_before) {
        delay_before_ = delay_before;
        return static_cast<B&>(*this);
    }

    // Set amount of time to wait in ms after each iteration.
    B& DelayAfter(uint16_t delay_after) {
        delay_after_ = delay_after;
        return static_cast<B&>(*this);
    }

    // Stop current effect and turn LED immeadiately off. Further calls to
    // Update() will have no effect.
    B& Stop() {
        Write(kZeroBrightness);
        state_ = ST_STOPPED;
        return static_cast<B&>(*this);
    }

    bool IsRunning() const { return state_ != ST_STOPPED; }

    // Reset to inital state
    B& Reset() {
        time_start_ = kTimeUndef;
        last_update_time_ = kTimeUndef;
        state_ = ST_RUNNING;
        return static_cast<B&>(*this);
    }

    // Sets the maximum brightness level. 255 is full brightness, 0 turns the
    // effect off. Currently, only upper 5 bits of the provided value are used
    // and stored.
    B& MaxBrightness(uint8_t level) {
        maxBrightness_ = level >> (8 - kBitsBrightness);
        return static_cast<B&>(*this);
    }

    // Returns current maximum brightness level. Since currently only upper 5
    // bits are used, lower 3 bits will always be 0.
    uint8_t MaxBrightness() const {
        return maxBrightness_ << (8 - kBitsBrightness);
    }

 protected:
    // update brightness of LED using the given brightness evaluator
    //  (brightness)                       ________________
    // on 255 |                         ¸-'
    //        |                      ¸-'
    //        |                   ¸-'
    // off 0  |________________¸-'
    //        |<-delay before->|<--period-->|<-delay after-> (time)
    //                         | func(t)    |
    //                         |<- num_repetitions times  ->
    bool Update(uint32_t now) {
        if (state_ == ST_STOPPED || !brightness_eval_) return false;

        // no need to process updates twice during one time tick.
        if (last_update_time_ == now) return true;

        if (last_update_time_ == kTimeUndef) {
            time_start_ = now + delay_before_;
        }
        last_update_time_ = now;

        if ((int32_t)(now - time_start_) < 0) return true;

        // t cycles in range [0..period+delay_after-1]
        const auto period = brightness_eval_->Period();
        const auto t = (now - time_start_) % (period + delay_after_);

        if (t < period) {
            state_ = ST_RUNNING;
            Write(brightness_eval_->Eval(t));
        } else {
            if (state_ == ST_RUNNING) {
                // when in delay after phase, just call Write()
                // once at the beginning.
                state_ = ST_IN_DELAY_AFTER_PHASE;
                Write(brightness_eval_->Eval(period - 1));
            }
        }

        if (IsForever()) return true;

        const auto time_end =
            time_start_ + (uint32_t)(period + delay_after_) * num_repetitions_ -
            1;

        if ((int32_t)(now - time_end) >= 0) {
            // make sure final value of t = (period-1) is set
            state_ = ST_STOPPED;
            Write(brightness_eval_->Eval(period - 1));
            return false;
        }
        return true;
    }

    B& SetBrightnessEval(BrightnessEvaluator* be) {
        brightness_eval_ = be;
        // start over after the brightness evaluator changed
        return Reset();
    }

 private:
    static constexpr uint8_t ST_STOPPED = 0;
    static constexpr uint8_t ST_RUNNING = 1;
    static constexpr uint8_t ST_IN_DELAY_AFTER_PHASE = 2;

    uint8_t state_ : 2;
    uint8_t bLowActive_ : 1;

    // Number of bits used to control brightness with MaxBrightness(). Using
    // only 5 bits here saves us a byte, since summing up with previous defs.
 public:
    static constexpr uint8_t kBitsBrightness = 5;
    static constexpr uint8_t kBrightnessStep = 1 << (8 - kBitsBrightness);

 private:
    uint8_t maxBrightness_ : kBitsBrightness;

    // this is where the BrightnessEvaluator object will be stored using
    // placment new.  Set MAX_SIZE to class occupying most memory
    static constexpr auto MAX_SIZE = sizeof(CandleBrightnessEvaluator);
    alignas(alignof(
        CloneableBrightnessEvaluator)) char brightness_eval_buf_[MAX_SIZE];

    static constexpr uint16_t kRepeatForever = 65535;
    uint16_t num_repetitions_ = 1;

    static constexpr uint32_t kTimeUndef = -1;
    uint32_t last_update_time_ = kTimeUndef;
    uint32_t time_start_ = kTimeUndef;

    uint16_t delay_before_ = 0;  // delay before the first effect starts
    uint16_t delay_after_ = 0;   // delay after each repetition
};

template<typename T> T* ptr(T &obj) {return &obj;}  // NOLINT
template<typename T> T* ptr(T *obj) {return obj;}

// a group of JLed objects which can be controlled simultanously, in parallel
// or sequentially.
template <typename T>
class TJLedSequence {
 protected:
    // update all leds parallel. Returns true while any of the JLeds is
    // active, else false
    bool UpdateParallel() {
        auto result = false;
        for (auto i = 0u; i < n_; i++) {
            result |= ptr(leds_[i])->Update();
        }
        return result;
    }

    // update all leds sequentially. Returns true while any of the JLeds is
    // active, else false
    bool UpdateSequentially() {
        if (cur_ >= n_) {
            return false;
        }
        if (!ptr(leds_[cur_])->Update()) {
            return ++cur_ < n_;
        }
        return true;;
    }

    void ResetLeds() {
        for (auto i = 0u; i < n_; i++) {
            ptr(leds_[i])->Reset();
        }
    }


 public:
    enum eMode { SEQUENCE, PARALLEL };
    TJLedSequence() = delete;

    template <size_t N>
    TJLedSequence(eMode mode, T (&leds)[N]) : TJLedSequence(mode, leds, N) {}

    TJLedSequence(eMode mode, T* leds, size_t n)
        : mode_{mode}, leds_{leds}, cur_{0}, n_{n} {}

    bool Update() {
        if (!is_running_) {
            return false;
        }

        const auto led_running = (mode_ == eMode::PARALLEL) ? UpdateParallel()
                                                  : UpdateSequentially();
        if (led_running) {
            return true;
        }

        // start next iteration of sequence
        cur_ = 0;
        ResetLeds();

        is_running_ = ++iteration_ < num_repetitions_  ||
                      num_repetitions_ == kRepeatForever;

        return is_running_;
    }

    void Reset() {
        ResetLeds();
        cur_ = 0;
        iteration_ = 0;
        is_running_ = true;
    }

    void Stop() {
        for (auto i = 0u; i < n_; i++) {
            leds_[i].Stop();
        }
    }

    // set number of repetitions for the sequence
    TJLedSequence<T> Repeat(uint16_t num_repetitions) {
        num_repetitions_ = num_repetitions;
        return *this;
    }

    // repeat Forever
    TJLedSequence<T> Forever() { return Repeat(kRepeatForever); }
    bool IsForever() const { return num_repetitions_ == kRepeatForever; }

 private:
    const eMode mode_;
    T* leds_;
    size_t cur_;
    const size_t n_;
    static constexpr uint16_t kRepeatForever = 65535;
    uint16_t num_repetitions_ = 1;
    uint16_t iteration_ = 0;
    bool is_running_ = true;
};

};  // namespace jled

#endif  // SRC_JLED_BASE_H_
